src/Core/Lib/BTreeDictionary.cs

```diff
@@ -0,0 +1,750 @@
+            Populate(entries);
+        }
+
+        private abstract class Node
+        {
+            public int count;       // # of direct children
+            public int totalCount;  // # of recursively reachable children.
+            public TKey[] keys;
+
+            public abstract (Node, Node) Add(TKey key, TValue value, BTreeDictionary<TKey, TValue> tree);
+
+            public abstract (TValue, bool) Get(TKey key, BTreeDictionary<TKey, TValue> tree);
+
+            public abstract (Node, Node) Set(TKey key, TValue value, BTreeDictionary<TKey, TValue> tree);
+
+            public abstract bool Remove(TKey key, BTreeDictionary<TKey, TValue> tree);
+
+            public override string ToString()
+            {
+                return $"{GetType().Name}: {count} items; keys: {string.Join(",",keys)}.";
+            }
+        }
+
+        /// <summary>
+        /// In a B+Tree, the values are held in the leaf nodes of the data structure.
+        /// </summary>
+        private class LeafNode : Node 
+        {
+            public LeafNode nextLeaf;   // leaves are threaded together for ease of enumeration.
+            public TValue[] values;
+
+            public LeafNode(int children)
+            {
+                this.keys = new TKey[children];
+                this.values = new TValue[children];
+            }
+
+            public override (Node, Node) Add(TKey key, TValue value, BTreeDictionary<TKey, TValue> tree)
+            {
+                int idx = Array.BinarySearch(keys, 0, count, key, tree.Comparer);
+                if (idx >= 0)
+                    throw new ArgumentException("Duplicate key.");
+                return Insert(~idx, key, value, tree);
+            }
+
+            public override (TValue, bool) Get(TKey key, BTreeDictionary<TKey, TValue> tree)
+            {
+                int idx = Array.BinarySearch(keys, 0, count, key, tree.Comparer);
+                if (idx < 0)
+                    return (default(TValue), false);
+                return (values[idx], true);
+            }
+
+            public override (Node, Node) Set(TKey key, TValue value, BTreeDictionary<TKey, TValue> tree)
+            {
+                int idx = Array.BinarySearch(keys, 0, count, key, tree.Comparer);
+                if (idx >= 0)
+                {
+                    values[idx] = value;
+                    return (this, null);
+                }
+                return Insert(~idx, key, value, tree);
+            }
+
+            public override bool Remove(TKey key, BTreeDictionary<TKey, TValue> tree)
+            {
+                int idx = Array.BinarySearch(keys, 0, count, key, tree.Comparer);
+                if (idx >= 0)
+                {
+                    --count;
+                    --totalCount;
+                    Array.Copy(keys, idx + 1, keys, idx, count - idx);
+                    Array.Copy(values, idx + 1, values, idx, count - idx);
```

Comment:
Could `idx + 1` be out of range?

Reply:
This is subtle but correct. Suppose there are two items  (`count` = 2) and we remove the item at `idx = 1`. We will decrement `count` to 1, and do `Array.Copy(keys, 1 + 1, keys, 1, 1 - 1)`. The item count of `1-1` makes the `Array.Copy` a no-op.